apps/src/blockly/addons/cdoSpritePointer.js

```diff
@@ -0,0 +1,120 @@
+// Helper functions for pointer block logic. In our pointer blocks (a.k.a.)
+// mini toolbox blocks) we "shadow" the image on the block from a source block.
+// This file contains functions to get and set the image on the pointer block.
+
+/**
+ * Get the image url for the given pointer block.
+ * @param {Block} block pointer block to get the image url for
+ * @param {*} pointerMetadataMap Object of the form {blockType: {imageSourceType: <string>, imageIndex: <number>}},
+ *  which maps pointer block types to the image source block type they can shadow and the index of the image to shadow.
+ * @param {string} imageSourceId Optional id of the image source block.
+ * @returns The url of the image that the pointer block should display, or an empty string if the block should display no image
+ */
+export function getPointerBlockImageUrl(
+  block,
+  pointerMetadataMap,
+  imageSourceId
+) {
+  const pointerData = pointerMetadataMap[block.type];
+  if (!pointerData || !block.inputList || block.inputList.length === 0) {
+    return '';
+  }
+  let imageSourceBlock = undefined;
+  const mainWorkspace = Blockly.getMainWorkspace();
+  if (imageSourceId !== undefined) {
+    imageSourceBlock = mainWorkspace.getBlockById(imageSourceId);
+  }
+  if (!imageSourceBlock) {
+    const rootBlock = block.getRootBlock();
+    if (rootBlock.type === pointerData.imageSourceType) {
+      imageSourceBlock = rootBlock;
+    } else if (rootBlock.id === block.id) {
+      const blockWorkspace = Blockly.Workspace.getById(block.workspace.id);
+      // If this block has itself as a root and is in a flyout workspace,
+      // it is a in a mini toolbox. A block can't be moved from one flyout to
+      // another, so if it's in a flyout, it is in its original flyout.
+      if (blockWorkspace && blockWorkspace.isFlyout && block.imageSourceId) {
+        imageSourceBlock = mainWorkspace.getBlockById(block.imageSourceId);
+      }
+    }
+  }
+  if (imageSourceBlock) {
+    return getImageUrlFromImageSource(imageSourceBlock, pointerData.imageIndex);
+  } else {
+    // The block is probably disconnected from any root or toolbox. Reset to
+    // default text.
+    return '';
```

Comment:
super-nit: Slight preference for default return statements to happen outside the else block, but feel free to take or leave!

Reply:
I like the clarity that we only return `''` if the `imageSourceBlock` is undefined.